source/common/upstream/load_balancer_impl.cc

```diff
@@ -1103,10 +1103,39 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
     // Cycle through hosts to achieve the intended offset behavior.
     // TODO(htuch): Consider how we can avoid biasing towards earlier hosts in the schedule across
     // refreshes for the weighted case.
-    if (!hosts.empty()) {
-      for (uint32_t i = 0; i < seed_ % hosts.size(); ++i) {
-        auto host =
-            scheduler.edf_->pickAndAdd([this](const Host& host) { return hostWeight(host); });
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.edf_lb_scheduler_init_fix")) {
+      // Only apply randomization if there are at least 2 hosts. If there's one host
+      // (or no hosts), its weight doesn't matter for the load-balancer.
+      if (hosts.size() > 1) {
+        // Compute the greatest common divisor of all the weights, to normalize
+        // the sum of weights to a smaller number. This can be avoided, but will
+        // result in higher numbers and more calls to `pickAndAdd()`.
+        const uint32_t host0_weight = static_cast<uint32_t>(std::ceil(hostWeight(*hosts[0])));
```

Comment:
wouldn't using `ceil` skew the proportion of host weights in some cases? E.g. we have host1 with weight 0.1 and host2 with weight 1 (proportion 1:10), by using ceil that proportion would turn into 1:1 and make host1 as equally likely to be picked as host2.

Reply:
Yeah, considering the additional factor like the slow start, the `hostWeight` may return a very minor value like 0.1.

And should we limit the max number of this initial randomize (or add a switch to it)? If weight difference of different endpoints is very very large, like 1 and 9999, then the initial randomize number will very large. cc @adisuissa

Reply:
I took that into consideration when looking for how to fix the bug.
I'm not aware of a way to compute a gcd of non-integers. The alternative is to keep the bug as is (which also impacts a case of 2 hosts one with weight=1, and the other with weight=0.1). If you know of an efficient  way to compute the length of the cycle with non-integer weights, please let me know.

The right way to fix this is probably convert this to static stride-scheduling ([grpc implementations](https://github.com/grpc/grpc-java/blob/fc03f2be9d3e0858d207015a5ff013a2bf03a77f/xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java#L425-L452)). However that will be a substantial change, and if added as an alternative, it will still not solve the current bug.

RE limiting the max number of randomized, I can probably add that somewhere (bootstrap?), but that seems more like a feature that will never go away to solve a bug.

Reply:
cc @adisuissa is there any input about the problem that @nezdolik raised. Basically, I think this is only point that may block this PR. (Ignore my suggestion about the max number of randomize, we can try to resolve it until it actually become a problem)